src/tools/rbd_mirror/group_replayer/BootstrapRequest.cc

```diff
@@ -0,0 +1,1446 @@
+    finish(r);
+    return;
+  }
+
+  create_local_mirror_group();
+}
+
+template <typename I>
+void BootstrapRequest<I>::create_local_mirror_group() {
+  dout(10) << dendl;
+
+  ceph_assert(
+      m_remote_mirror_group.state == cls::rbd::MIRROR_GROUP_STATE_ENABLED &&
+      m_remote_mirror_group_primary);
+
+  librados::ObjectWriteOperation op;
+  m_local_mirror_group = {m_global_group_id,
+                          m_remote_mirror_group.mirror_image_mode,
+                          cls::rbd::MIRROR_GROUP_STATE_ENABLED};
+  librbd::cls_client::mirror_group_set(&op, *m_local_group_id,
+                                       m_local_mirror_group);
+  auto comp = create_rados_callback<
+      BootstrapRequest<I>,
+      &BootstrapRequest<I>::handle_create_local_mirror_group>(this);
+
+  int r = m_local_io_ctx.aio_operate(RBD_MIRRORING, comp, &op);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void BootstrapRequest<I>::handle_create_local_mirror_group(int r) {
+  dout(10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    derr << "error creating local mirror group: " << cpp_strerror(r) << dendl;
+    finish(r);
+    return;
+  }
+
+  create_local_non_primary_group_snapshot();
+}
+
+template <typename I>
+void BootstrapRequest<I>::create_local_non_primary_group_snapshot() {
+  dout(10) << dendl;
+
+  RemotePoolMeta remote_pool_meta;
+  int r = m_pool_meta_cache->get_remote_pool_meta(m_remote_io_ctx.get_id(),
+                                                  &remote_pool_meta);
+  if (r < 0 || remote_pool_meta.mirror_peer_uuid.empty()) {
+    derr << "failed to retrieve mirror peer uuid from remote image pool"
+         << dendl;
+    finish(r < 0 ? r : -EINVAL);
+    return;
+  }
+
+  librados::ObjectWriteOperation op;
+  std::string group_snap_id = librbd::util::generate_image_id(m_local_io_ctx);
+  cls::rbd::GroupSnapshot group_snap{
+    group_snap_id,
+    cls::rbd::MirrorGroupSnapshotNamespace{
+      cls::rbd::MIRROR_SNAPSHOT_STATE_NON_PRIMARY,
+      {}, remote_pool_meta.mirror_peer_uuid, {}},
+      prepare_non_primary_mirror_snap_name(m_global_group_id, group_snap_id),
+      cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE};
+  librbd::cls_client::group_snap_set(&op, group_snap);
+  group_snap.state = cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE;
+  librbd::cls_client::group_snap_set(&op, group_snap);
+
+  auto comp = create_rados_callback<
+      BootstrapRequest<I>,
+      &BootstrapRequest<I>::handle_create_local_non_primary_group_snapshot>(this);
+
+  r = m_local_io_ctx.aio_operate(
+      librbd::util::group_header_name(*m_local_group_id), comp, &op);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void BootstrapRequest<I>::handle_create_local_non_primary_group_snapshot(int r) {
+  dout(10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    derr << "error creating local non-primary group snapshot: "
+         << cpp_strerror(r) << dendl;
+    finish(r);
+    return;
+  }
```

Comment:
@trociny Do you remember why we need this  initial snapshot on the non-primary? To keep it aligned with the image snapshot mirroring want to avoid/remove this snap creation.

At-least I cannot discover a valid reason from my read. Thoughts? 

Thanks!

Reply:
cc: @idryomov @nbalacha

Reply:
Removed it for now.